### PR TITLE
fix append! for SubArray

### PR DESCRIPTION
**Pull request by bkamins**
Changed `append!` to use `CatArrOrSub`. I also changed the use of `length` to avoid `refs`.
@nalimilan Did you have any specific reason to use it directly?

### CODE REVIEW

File: test/11_array.jl [hunk 1]
```diff
@@ -219,10 +219,17 @@ using CategoricalArrays: DefaultRefType, catvaluetype, leveltype
     @test length(x) == 15
     @test x == ["c", "b", "b", "a", "zz", "c", "c", "b", "b", "a", "zz", "c", "z", "y", "x"]
     @test levels(x) == ["e", "a", "b", "c", "zz", "x", "y", "z"]
+    z1 = view(CategoricalVector{String, R}(["ex1", "ex2"]), 1)
```
[the remark this answers is not in the document]
Comment (bkamins) [reply test/11_array.jl]: I moved the tests as you have requested (and I admit that the testing code is really hard to work with). However, I retained the tests of full range of categorical arrays as in the future - when we fix `append!` to property alter ordering they should be performed.

File: test/13_arraycommon.jl [hunk 2]
```diff
@@ -979,4 +979,121 @@ end
     @test x == [3, 3, 3]
 end
 
+@testset "append! ordered=$ordered reftype=$R" for ordered in (false, true),
+    R in (CategoricalArrays.DefaultRefType, UInt8, UInt, Int8, Int)
```
[the remark this answers is not in the document]
Comment (bkamins) [reply test/13_arraycommon.jl]: But does it hurt? I would retain it.

**Comment by bkamins** [comment]
This should be good for a review now.

I have added one test to broadcast, as we did not have a coverage of `CategoricalString` type there.

**Comment by bkamins** [comment]
I am not sure why codecov went down (I cannot view the results) - but it should actually go up as I have added not removed tests.

File: test/13_arraycommon.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (nalimilan) [reply test/13_arraycommon.jl]: It doesn't really hurt but it makes the run take longer, and there's really no reason why it would work for one `R` type and not another. If you really want to test corner cases with `R`, I'd rather check that `append!` works when the inputs have different type parameters -- but even that sounds obvious given how the function is implemented.

[the remark this answers is not in the document]
Comment (bkamins) [reply test/13_arraycommon.jl]: OK - I have removed `R` from tests.

**Comment by nalimilan** [comment]
Thanks! Coverage is broken these days.

I've tagged https://github.com/JuliaLang/METADATA.jl/pull/19647.
